47-backend-frontend/proposal.md

```diff
@@ -0,0 +1,65 @@
+
+I was looking through `atc/atccmd/command.go` and realized that the things I was looking for ("periodic asynchronous procedures") all fit more precisely into the category of `github.com/tedsuo/ifrit/grouper.Member` values.
+
+https://github.com/concourse/concourse/blob/release/5.5.x/atc/atccmd/command.go#L796-L890
+
+I started frantically tabulating each of these `grouper.Member`s and the intervals on which they ran, hoping to find something that might possibly cohere with my colleague's observation. Unfortunately it came to nothing, but the plot thickens.
+
+Eventually we got some heap dumps using the pprof endpoint:
+
+https://github.com/concourse/concourse/blob/release/5.5.x/atc/atccmd/command.go#L11
+
+and then we discovered that the `ListAllJobs` endpoint was ultimately mostly responsible for the memory consumption, and we made a PR adding a config flag that to disable specifically that badly-behaved endpoint: https://github.com/concourse/concourse/pull/5340
+
+## A request for non-invasive operator-configurable observability
+
+How can we improve situations like this in the future? I would like the following acceptance criteria:
+
+A way to report compute-resource consumption of Concourse by "member" (or smaller), such that
+* data can be harvested in real-time
+* operators can configure how they wish to group the members
+* operators can opt-in or -out of each member's functionality, at a level of granularity that permits disabling a single endpoint.
+
+Probably the existing mechanism of Linux processes and all of the existing technology for observing them should be leveraged to the greatest extent possible - we can always add our own advisory/application-level monitoring but it's probably always going to be chasing old incidents and won't be able to give adequate information about real-time production scenarios.
+
+I can imagine a world where a common way to introduce Concourse features is by adding or removing "members" of this kind - frankly, `lidar`, the big feature in v5.6.0, was basically a drop-in replacement for the existing `radar` that performed resource checks.
+
+I think there's a lot of value in placing control over these things in the hands of the operator, especially when operators may have strict change management controls that forbid them from modifying the actual software in production, but may be more lenient about making configuration changes.
+
+# Proposal
+
+Broadly, we need a solution that lets operators choose whether they want to run the various Concourse computations in `ifrit/grouper.Members` or actual separate unix processes -- these things are supposed to be nearly-interchangeable concepts anyway. We could provide this configurability at many levels of granularity, so let's focus this RFC on a simple case.
```

Comment:
I'm struggling a bit with understanding if I'm looking at the RFC with the right angle :grimacing: 

e.g., if I take this exactly sentence, and append to it a "so it solves what I'm trying to solve", is the following correct?

> [...] we need a solution that lets operators choose whether they want to run the various Concourse computations in `ifrit/grouper.Members` or actual separate unix processes, **so that observability of the web node is improved, being even able to know which API endpoint is eating all my memory without pulling a heap dump** 

(just trying to make sure I'm with the right mindset)

Reply:
I've definitely made a mess here - as though talking about microservices didn't give it away. I should maybe restate my goal as "figure out WTF is eating my memory without a heap dump". In this case the RFC just proposes a way to figure out whether the frontend or backend is to blame.

Reply:
The key thing that is also missing is the ~3h offline conversation between @pivotal-jwinters and I where these goals were heavily revised...